### PR TITLE
[codex] add per-env trainer metrics

### PR DESCRIPTION
**Pull request by samsja**
## Summary

Adds environment metadata to the orchestrator-to-trainer batch path so trainer-side token metrics can be broken down by environment in W&B.

- Carries required `env_name` on `TrainingSample` and expands it into required per-token `env_names` on `MicroBatch`.
- Validates per-token env-name length when loading micro-batches into trainer tensors, then keeps the training loop non-defensive.
- Reuses the existing `Tensors` accumulator for grouped trainer token stats, logging per-env `entropy` and `mismatch_kl` as `{metric}/{env}/{mean,std,max}` while preserving existing aggregate metrics like `entropy/mean`.
- Makes `Tensors.compute_stats()` robust to sparse dynamic keys across ranks, so per-env keys cannot desynchronize distributed collectives.
- Computes importance-ratio/mismatch-KL once per microbatch, passes those tensors through the loss path and per-env logging, derives the DPPO probability delta from the shared ratio, and reuses the already-copied entropy tensor for env splits.
- Updates focused tests plus the monitoring skill docs.

## Validation

- `uv run pytest tests/unit/orchestrator/test_batch.py tests/unit/orchestrator/test_teacher_logprobs.py tests/unit/orchestrator/test_sft_trajectories.py tests/unit/orchestrator/test_trajectories.py tests/unit/train/rl/test_packer.py`
- `PYTEST_OUTPUT_DIR=/tmp/outputs uv run pytest $(git ls-files 'tests/unit/**/test_*.py' 'tests/unit/test_*.py') -m "not gpu"` (`344 passed, 65 deselected`)
- `uv run pytest tests/unit/train/rl/test_loss.py -m gpu` (`6 passed`)
- `uv run ruff check src/prime_rl/trainer/rl/loss.py src/prime_rl/trainer/rl/train.py`
- `uv run ruff format --check src/prime_rl/trainer/rl/loss.py src/prime_rl/trainer/rl/train.py`
- Reverse-text W&B smoke run: trainer https://wandb.ai/primeintellect/reverse-text/runs/07xfclyq, orchestrator https://wandb.ai/primeintellect/reverse-text/runs/ytp92whl

---

> [!NOTE]
> **Medium Risk**
> Changes the orchestrator→trainer transport schema and distributed metric aggregation to carry per-token environment names and log per-env trainer stats; mistakes here could break batch serialization or desynchronize multi-rank metric collectives.
> 
> **Overview**
> Adds environment metadata to the training data path so trainer-side token metrics can be reported *per environment*.
> 
> `TrainingSample` now requires `env_name`, which is propagated through the orchestrator and expanded into per-token `env_names` on `MicroBatch` (including packing/truncation/padding validation). The trainer loop then splits token-level `entropy` and `mismatch_kl` by env and logs keys as `entropy/{all,env}/...` and `mismatch_kl/{all,env}/...`.
> 
> Makes `Tensors.compute_stats()` robust to dynamic/sparse metric keys across distributed ranks (gather keys first, handle empty tensors), updates W&B metric filtering to allow env-scoped stats, factors mismatch-KL computation into `compute_importance_ratio_and_mismatch_kl`, and updates tests plus monitoring docs to reflect the new metric keys.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 09c793065514890bd846a5cdc7f99e8aa801d1cd. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/trainer/rl/train.py [hunk 1]
```diff
@@ -477,6 +479,17 @@ def load_run_checkpoint(_optimizer, idx: int) -> None:
...
+            env_names = micro_batch.get("env_names")
+            if env_names is not None:
+                masked_env_names = [env_name for env_name, keep in zip(env_names, loss_mask.flatten().tolist()) if keep]
+                env_tensors.append("entropy", out["entropy"][loss_mask].detach().to("cpu"), masked_env_names)
+                if "mismatch_kl" in loss_tensors:
+                    with torch.no_grad():
+                        log_importance_ratio = out["logprobs"] - inference_logprobs
+                        importance_ratio = torch.exp(log_importance_ratio)
+                        mismatch_kl = importance_ratio - log_importance_ratio - 1
+                    env_tensors.append("mismatch_kl", mismatch_kl[loss_mask].detach().to("cpu"), masked_env_names)
```
Comment (samsja) [inline src/prime_rl/trainer/rl/train.py]: env name shouldn't be allowed to be None or not to be here

File: skills/monitor-run/SKILL.md [hunk 2]
```diff
@@ -145,6 +145,8 @@ These tell you whether training is healthy or diverging.
 |--------|--------|-------------|
 | `mismatch_kl/mean` | trainer | KL divergence between trainer and (old) inference policy  |
 | `entropy/mean` | trainer | policy entropy |
+| `mismatch_kl/{env}/{mean,std,max}` | trainer | per-environment KL mismatch over trainable tokens (W&B) |
```
Comment (mikasenghaas) [inline skills/monitor-run/SKILL.md]: we should support `mismatch_kl/{all,env}/{mean,std,max}` like on orch, same for all other per-env metrics

File: src/prime_rl/orchestrator/trajectories.py [hunk 3]
```diff
@@ -337,6 +337,7 @@ def make_sample(tokens: dict[str, Any]) -> TrainingSample:
             completion_temperatures=[temperature] * len(completion_ids),
             teacher_logprobs=None,
             advantage=None,
+            env_name=output.get("env_name") or "",
```
Comment (mikasenghaas) [inline src/prime_rl/orchestrator/trajectories.py]: don't think we need/ should be defensive here

File: src/prime_rl/trainer/rl/data.py [hunk 4]
```diff
@@ -192,6 +195,13 @@ def get_batch(self) -> list[TensorMicroBatch]:
 
     def _micro_batch_to_tensor(self, micro_batch: MicroBatch) -> TensorMicroBatch:
         """Convert a MicroBatch (msgspec struct with lists) to a TensorMicroBatch (dict with tensors)."""
+        if len(micro_batch.env_names) != len(micro_batch.input_ids):
+            raise ValueError(
+                f"MicroBatch.env_names must match input_ids length: "
+                f"env_names={len(micro_batch.env_names)}, input_ids={len(micro_batch.input_ids)}"
+            )
+        if any(not env_name for env_name, keep in zip(micro_batch.env_names, micro_batch.loss_mask) if keep):
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/rl/data.py]: also not sure we need this assertion here?

File: src/prime_rl/trainer/rl/train.py [hunk 5]
```diff
@@ -477,6 +479,24 @@ def load_run_checkpoint(_optimizer, idx: int) -> None:
...
 
+            env_names = micro_batch["env_names"]
+            loss_mask_values = loss_mask.flatten().tolist()
+            if len(env_names) != len(loss_mask_values):
+                raise ValueError(
+                    f"micro_batch env_names length must match loss_mask length: "
+                    f"env_names={len(env_names)}, loss_mask={len(loss_mask_values)}"
+                )
+            masked_env_names = [env_name for env_name, keep in zip(env_names, loss_mask_values) if keep]
+            if any(not env_name for env_name in masked_env_names):
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/rl/train.py]: also here overly defensive?

File: src/prime_rl/trainer/rl/train.py [hunk 6]
```diff
@@ -477,6 +479,24 @@ def load_run_checkpoint(_optimizer, idx: int) -> None:
...
+                raise ValueError("micro_batch env_names must be set for every trainable token")
+            env_tensors.append("entropy", out["entropy"][loss_mask].detach().to("cpu"), masked_env_names)
+            if "mismatch_kl" in loss_tensors:
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/rl/train.py]: why this conditional? do we ever not log mismatch?

File: src/prime_rl/trainer/rl/train.py [hunk 7]
```diff
@@ -477,6 +479,24 @@ def load_run_checkpoint(_optimizer, idx: int) -> None:
...
+                with torch.no_grad():
+                    log_importance_ratio = out["logprobs"] - inference_logprobs
+                    importance_ratio = torch.exp(log_importance_ratio)
+                    mismatch_kl = importance_ratio - log_importance_ratio - 1
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/rl/train.py]: should prob make this a util

File: src/prime_rl/trainer/batch.py [hunk 8]
```diff
@@ -14,6 +14,9 @@ def prepare_sample(training_example: TrainingSample, seq_len: int) -> MicroBatch
     advantages = [training_example.advantage] * len(input_ids)
     position_ids = list(range(len(input_ids)))
     mm_token_type_ids = training_example.mm_token_type_ids
+    if not training_example.env_name:
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/batch.py]: can remove

File: src/prime_rl/trainer/utils.py [hunk 9]
```diff
@@ -382,6 +382,64 @@ def compute_stats(self) -> dict[str, float | int]:
         return metrics
 
 
+class GroupedTensors:
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/utils.py]: could consider doing a `dict[str, Tensors]` which is keyed by env to have one less abstraction. but ig requires some bookkepping around which token belongs to which env

File: src/prime_rl/trainer/rl/train.py [hunk 10]
```diff
@@ -477,6 +479,24 @@ def load_run_checkpoint(_optimizer, idx: int) -> None:
             tensors["entropy"].append(out["entropy"][loss_mask].detach().to("cpu"))
             tensors["loss"].append(loss.detach().to("cpu").unsqueeze(0))
 
+            env_names = micro_batch["env_names"]
+            loss_mask_values = loss_mask.flatten().tolist()
+            if len(env_names) != len(loss_mask_values):
+                raise ValueError(
+                    f"micro_batch env_names length must match loss_mask length: "
+                    f"env_names={len(env_names)}, loss_mask={len(loss_mask_values)}"
+                )
```
Comment (samsja) [inline src/prime_rl/trainer/rl/train.py]: we already check this

File: src/prime_rl/trainer/rl/train.py (hunk 5 above)
Comment (samsja) [inline src/prime_rl/trainer/rl/train.py]: no need for check

File: src/prime_rl/trainer/rl/train.py (hunk 6 above)
Comment (samsja) [inline src/prime_rl/trainer/rl/train.py]: why ? this ia alwaus present and if it wasn't we should fail, like lets just remove the if and keep like before

File: src/prime_rl/transport/types.py [hunk 11]
```diff
@@ -27,6 +27,7 @@ class TrainingSample(msgspec.Struct, array_like=True, gc=False, omit_defaults=Tr
     mm_token_type_ids: list[int] | None = None
 
     sft_loss: bool = False  # When True, trainer uses SFT loss instead of RL loss for this sample
+    env_name: str = ""
```
Comment (mikasenghaas) [inline src/prime_rl/transport/types.py]: would not put a default here to ensure we fail if this is not set

File: src/prime_rl/trainer/rl/loss.py [hunk 12]
```diff
@@ -136,7 +142,7 @@ def default_loss_fn(inputs: LossInputs, loss_config: DefaultLossConfig) -> LossO
 
     log_importance_ratio = trainer_logprobs - inference_logprobs
     importance_ratio = torch.exp(log_importance_ratio)
-    mismatch_kl = importance_ratio - log_importance_ratio - 1
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/rl/loss.py:142]: ah ig we haver some redundant computation here? is this negligble?

File: src/prime_rl/trainer/rl/train.py [hunk 13]
```diff
@@ -477,6 +478,24 @@ def load_run_checkpoint(_optimizer, idx: int) -> None:
...
             tensors["loss"].append(loss.detach().to("cpu").unsqueeze(0))
 
+            env_names = micro_batch["env_names"]
+            masked_env_names = [env_name for env_name, keep in zip(env_names, loss_mask.flatten().tolist()) if keep]
+            env_to_indices: dict[str, list[int]] = {}
+            for idx, env_name in enumerate(masked_env_names):
+                env_to_indices.setdefault(env_name, []).append(idx)
+
+            entropy = out["entropy"][loss_mask].detach().to("cpu")
+            tensors["entropy/all"].append(entropy)
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/rl/train.py]: wait now we have `entropy/all/...` and `entropy/...`?

File: src/prime_rl/trainer/rl/loss.py [hunk 14]
```diff
@@ -19,6 +19,9 @@ class LossInputs:
     teacher_logprobs: Float[Tensor, " seq"] | None
     advantages: Float[Tensor, " seq"]
     loss_mask: Bool[Tensor, " seq"]
+    log_importance_ratio: Float[Tensor, " seq"] | None = None
+    importance_ratio: Float[Tensor, " seq"] | None = None
+    mismatch_kl: Float[Tensor, " seq"] | None = None
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/rl/loss.py]: on second thought, would prefer to not pass those but compute inside the loss functions ince those are functions of train+infer logprobs. if the orch also computes thme for logging, so be it. should be cheap enough

File: src/prime_rl/trainer/rl/loss.py [hunk 15]
```diff
@@ -122,9 +131,12 @@ def default_loss_fn(inputs: LossInputs, loss_config: DefaultLossConfig) -> LossO
...
     loss_mask = inputs.loss_mask
 
-    trainer_probs = torch.exp(trainer_logprobs)
+    log_importance_ratio, importance_ratio, mismatch_kl = compute_importance_ratio_and_mismatch_kl(
+        trainer_logprobs, inference_logprobs
+    )
+
     inference_probs = torch.exp(inference_logprobs)
-    probs_diff = trainer_probs - inference_probs
+    probs_diff = inference_probs * (importance_ratio - 1)
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/rl/loss.py]: this is a wild way to compute a difference lol

File: src/prime_rl/trainer/rl/train.py [hunk 16]
```diff
@@ -486,12 +504,14 @@ def load_run_checkpoint(_optimizer, idx: int) -> None:
             # Add loss tensors to tensor dict for logging purposes
             for key, loss_tensor in loss_tensors.items():
                 loss_tensor = loss_tensor.detach().to("cpu")
+                if key == "mismatch_kl":
+                    key = "mismatch_kl/all"
```
[the remark this answers is not in the document]
Comment (mikasenghaas) [reply src/prime_rl/trainer/rl/train.py]: damn i wasn't aware we were not doing per-token avg for our loss metrics -- we should prob change this? @faresobeid

File: src/prime_rl/trainer/rl/train.py [hunk 17]
```diff
@@ -486,12 +504,14 @@ def load_run_checkpoint(_optimizer, idx: int) -> None:
             # Add loss tensors to tensor dict for logging purposes
             for key, loss_tensor in loss_tensors.items():
                 loss_tensor = loss_tensor.detach().to("cpu")
+                if key == "mismatch_kl":
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/rl/train.py]: should we not just compute mismatch kl in the main trainer loop and export it as a loss metric (we may still compute it there but we just dont log from it). i think this makes semantic sense bc we want to report mismatch even if our loss doens't depend on it

File: src/prime_rl/trainer/rl/loss.py (hunk 15 above)
Comment (samsja) [reply src/prime_rl/trainer/rl/loss.py]: bru what

File: src/prime_rl/trainer/rl/train.py (hunk 17 above)
Comment (samsja) [reply src/prime_rl/trainer/rl/train.py]: sure

**Review by mikasenghaas — APPROVED** [review]
